File: mobile/android/FamilyPilotSpeechWebBridgeV1.kt
package com.familypilot.voice

import android.Manifest
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import android.webkit.WebView
import androidx.core.content.ContextCompat
import androidx.webkit.JavaScriptReplyProxy
import androidx.webkit.WebMessageCompat
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import org.json.JSONObject

/** Origin-scoped WebView bridge for a button-controlled on-device speech session. */
class FamilyPilotSpeechWebBridgeV1(
    private val webView: WebView,
    private val speech: FamilyPilotOnDeviceSpeechV1,
    private val allowedOriginRules: Set<String>,
    private val requestMicrophonePermission: (((Boolean) -> Unit) -> Unit),
) {
    companion object {
        private const val JS_OBJECT = "FamilyPilotNativeSpeechAndroidBridgeV1"
        private const val ENTRY = "./familypilot-voice-v1-native-entry.js"
        private val ENTRY_BOOTSTRAP = """
            (()=>{const load=()=>{if(document.querySelector('script[data-fp-voice-native-entry]'))return;const s=document.createElement('script');s.src='$ENTRY';s.dataset.fpVoiceNativeEntry='1';document.head.appendChild(s);};if(document.readyState==='loading')document.addEventListener('DOMContentLoaded',load,{once:true});else load();})();
        """.trimIndent()
    }

    private data class Session(
        val id: String,
        val replyProxy: JavaScriptReplyProxy,
        val chunks: MutableList<String> = mutableListOf(),
        var partial: String = "",
        var stopping: Boolean = false,
    )

    private val mainHandler = Handler(Looper.getMainLooper())
    private var installed = false
    private var session: Session? = null

    fun install(callback: (Boolean) -> Unit) {
        mainHandler.post {
            if (installed) { callback(true); return@post }
            if (allowedOriginRules.isEmpty() || allowedOriginRules.contains("*")) { callback(false); return@post }
            if (!WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER) || !WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) { callback(false); return@post }
            WebViewCompat.addWebMessageListener(webView, JS_OBJECT, allowedOriginRules, WebViewCompat.WebMessageListener { _, message, _, isMainFrame, replyProxy ->
                if (!isMainFrame) { reply(replyProxy, "", mapOf("ok" to false, "error" to "native_speech_bridge_frame_rejected")); return@WebMessageListener }
                handle(message, replyProxy)
            })
            WebViewCompat.addDocumentStartJavaScript(webView, ENTRY_BOOTSTRAP, allowedOriginRules)
            installed = true
            callback(true)
        }
    }

    private fun handle(message: WebMessageCompat, replyProxy: JavaScriptReplyProxy) {
        val raw = message.data ?: return
        val request = try { JSONObject(raw) } catch (_: Exception) { reply(replyProxy, "", mapOf("ok" to false, "error" to "native_speech_bridge_invalid_request")); return }
        val id = request.optString("id")
        when (request.optString("action")) {
            "isAvailable" -> speech.checkAvailability { availability ->
                val available = when (availability) {
                    FamilyPilotOnDeviceSpeechV1.Availability.READY,
                    FamilyPilotOnDeviceSpeechV1.Availability.MICROPHONE_PERMISSION_REQUIRED,
                    FamilyPilotOnDeviceSpeechV1.Availability.LANGUAGE_SUPPORT_UNKNOWN -> true
                    else -> false
                }
                reply(replyProxy, id, mapOf("available" to available, "status" to availability.name.lowercase()))
            }
            "recognize" -> beginSession(id, replyProxy)
            "stop" -> stopSession(id, replyProxy)
            else -> reply(replyProxy, id, mapOf("ok" to false, "error" to "native_speech_bridge_unknown_action"))
        }
    }

    private fun beginSession(id: String, replyProxy: JavaScriptReplyProxy) {
        if (session != null) { reply(replyProxy, id, mapOf("ok" to false, "error" to "recognition_busy")); return }
        val hasPermission = ContextCompat.checkSelfPermission(webView.context, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED
        if (hasPermission) { startSession(id, replyProxy); return }
        mainHandler.post { requestMicrophonePermission { granted -> if (granted) startSession(id, replyProxy) else reply(replyProxy, id, mapOf("ok" to false, "error" to "microphone_permission_denied")) } }
    }

    private fun startSession(id: String, replyProxy: JavaScriptReplyProxy) {
        if (session != null) { reply(replyProxy, id, mapOf("ok" to false, "error" to "recognition_busy")); return }
        val current = Session(id, replyProxy)
        session = current
        startSegment(current)
    }

    private fun startSegment(current: Session) {
        if (session !== current || current.stopping) return
        current.partial = ""
        speech.recognize(
            callback = { result ->
                mainHandler.post {
                    if (session !== current) return@post
                    when (result) {
                        is FamilyPilotOnDeviceSpeechV1.Result.Success -> {
                            if (result.text.isNotBlank()) current.chunks.add(result.text.trim())
                            current.partial = ""
                            if (current.stopping) complete(current) else startSegment(current)
                        }
                        is FamilyPilotOnDeviceSpeechV1.Result.Failure -> {
                            if (current.stopping) {
                                promotePartial(current)
                                complete(current)
                            } else if (isTransient(result.reason)) {
                                promotePartial(current)
                                mainHandler.postDelayed({ startSegment(current) }, 120)
                            } else fail(current, result.reason)
                        }
                    }
                }
            },
            onPartial = { text ->
                mainHandler.post {
                    if (session !== current || current.stopping) return@post
                    val trimmed = text.trim()
                    if (trimmed.isNotEmpty()) {
                        current.partial = trimmed
                        publishPartial(current)
                    }
                }
            },
        )
    }

    private fun publishPartial(current: Session) {
        if (session !== current || current.stopping) return
        val text = (current.chunks + listOf(current.partial)).filter { it.isNotBlank() }.joinToString(" ").trim()
        if (text.isNotEmpty()) reply(current.replyProxy, current.id, mapOf("event" to "partial", "text" to text))
    }

    private fun promotePartial(current: Session) {
        val text = current.partial.trim()
        if (text.isNotEmpty()) current.chunks.add(text)
        current.partial = ""
    }

    private fun stopSession(id: String, replyProxy: JavaScriptReplyProxy) {
        val current = session
        if (current == null) { reply(replyProxy, id, mapOf("ok" to false, "error" to "recognition_not_active")); return }
        current.stopping = true
        speech.stopListening()
        reply(replyProxy, id, mapOf("ok" to true, "stopping" to true))
        mainHandler.postDelayed({
            if (session === current) {
                speech.cancel()
                promotePartial(current)
                if (current.chunks.isNotEmpty()) complete(current) else fail(current, "empty_transcript")
            }
        }, 5000)
    }

    private fun complete(current: Session) {
        if (session !== current) return
        session = null
        val text = current.chunks.joinToString(" ").trim()
        if (text.isEmpty()) reply(current.replyProxy, current.id, mapOf("ok" to false, "error" to "empty_transcript"))
        else reply(current.replyProxy, current.id, mapOf("ok" to true, "text" to text))
    }

    private fun fail(current: Session, reason: String) {
        if (session !== current) return
        session = null
        speech.cancel()
        reply(current.replyProxy, current.id, mapOf("ok" to false, "error" to reason))
    }

    private fun isTransient(reason: String): Boolean =
        reason == "empty_transcript" || reason == "speech_recognition_failed:6" || reason == "speech_recognition_failed:7"

    private fun reply(replyProxy: JavaScriptReplyProxy, id: String, values: Map<String, Any>) {
        val payload = JSONObject(); payload.put("id", id); values.forEach { (key, value) -> payload.put(key, value) }
        mainHandler.post { replyProxy.postMessage(payload.toString()) }
    }
}

File: mobile/ios/FamilyPilotSpeechWebBridgeV1.swift
import Foundation
import WebKit

/// Main-frame-only bridge from trusted FamilyPilot web content to a button-controlled on-device speech session.
/// Async transcript replies remain pinned to the frame/security origin that started the session.
final class FamilyPilotSpeechWebBridgeV1: NSObject, WKScriptMessageHandler {
    static let handlerName = "FamilyPilotNativeSpeechIOSBridgeV1"
    private static let entryBootstrap = """
    (()=>{
      const load=()=>{
        if(document.querySelector('script[data-fp-voice-native-entry]'))return;
        const s=document.createElement('script');
        s.src='./familypilot-voice-v1-native-entry.js';
        s.dataset.fpVoiceNativeEntry='1';
        document.head.appendChild(s);
      };
      if(document.readyState==='loading')document.addEventListener('DOMContentLoaded',load,{once:true});else load();
    })();
    """

    private final class Session {
        let id: String
        let frame: WKFrameInfo
        var chunks: [String] = []
        var partial = ""
        var stopping = false

        init(id: String, frame: WKFrameInfo) {
            self.id = id
            self.frame = frame
        }
    }

    private let speech: FamilyPilotOnDeviceSpeechV1
    private let allowedSchemes: Set<String>
    private let allowedHosts: Set<String>
    private weak var webView: WKWebView?
    private var session: Session?

    init(
        speech: FamilyPilotOnDeviceSpeechV1,
        allowedSchemes: Set<String>,
        allowedHosts: Set<String>
    ) {
        self.speech = speech
        self.allowedSchemes = allowedSchemes
        self.allowedHosts = allowedHosts
    }

    func install(on configuration: WKWebViewConfiguration) -> Bool {
        guard !allowedSchemes.isEmpty else { return false }
        configuration.userContentController.add(self, name: Self.handlerName)
        configuration.userContentController.addUserScript(
            WKUserScript(source: Self.entryBootstrap, injectionTime: .atDocumentStart, forMainFrameOnly: true)
        )
        return true
    }

    func bind(webView: WKWebView) {
        self.webView = webView
    }

    func uninstall(from configuration: WKWebViewConfiguration) {
        configuration.userContentController.removeScriptMessageHandler(forName: Self.handlerName)
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        let frame = message.frameInfo
        guard frame.isMainFrame, isAllowed(frame.securityOrigin) else { return }
        guard let body = message.body as? [String: Any],
              let id = body["id"] as? String,
              let action = body["action"] as? String else {
            respond(id: "", values: ["ok": false, "error": "native_speech_bridge_invalid_request"], frame: frame)
            return
        }

        switch action {
        case "isAvailable":
            respond(id: id, values: ["available": speech.isAvailable()], frame: frame)
        case "recognize":
            beginSession(id: id, frame: frame)
        case "stop":
            stopSession(stopRequestID: id, frame: frame)
        default:
            respond(id: id, values: ["ok": false, "error": "native_speech_bridge_unknown_action"], frame: frame)
        }
    }

    private func beginSession(id: String, frame: WKFrameInfo) {
        guard session == nil else {
            respond(id: id, values: ["ok": false, "error": "recognition_busy"], frame: frame)
            return
        }
        let current = Session(id: id, frame: frame)
        session = current
        startSegment(current)
    }

    private func startSegment(_ current: Session) {
        guard session === current, !current.stopping else { return }
        current.partial = ""
        speech.recognize(
            onPartial: { [weak self, weak current] text in
                guard let self, let current else { return }
                DispatchQueue.main.async {
                    guard self.session === current, !current.stopping else { return }
                    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
                    guard !trimmed.isEmpty else { return }
                    current.partial = trimmed
                    self.publishPartial(current)
                }
            },
            completion: { [weak self, weak current] result in
                guard let self, let current else { return }
                DispatchQueue.main.async {
                    guard self.session === current else { return }
                    switch result {
                    case .success(let text):
                        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
                        if !trimmed.isEmpty { current.chunks.append(trimmed) }
                        current.partial = ""
                        if current.stopping { self.complete(current) }
                        else { self.startSegment(current) }
                    case .failure(let error):
                        if current.stopping {
                            self.promotePartial(current)
                            self.complete(current)
                        } else if error == .emptyTranscript || error == .recognitionFailed {
                            self.promotePartial(current)
                            DispatchQueue.main.asyncAfter(deadline: .now() + 0.12) { [weak self, weak current] in
                                guard let self, let current else { return }
                                self.startSegment(current)
                            }
                        } else {
                            self.fail(current, error: Self.errorCode(error))
                        }
                    }
                }
            }
        )
    }

    private func publishPartial(_ current: Session) {
        guard session === current, !current.stopping else { return }
        let text = (current.chunks + [current.partial])
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        respond(id: current.id, values: ["event": "partial", "text": text], frame: current.frame)
    }

    private func promotePartial(_ current: Session) {
        let text = current.partial.trimmingCharacters(in: .whitespacesAndNewlines)
        if !text.isEmpty { current.chunks.append(text) }
        current.partial = ""
    }

    private func stopSession(stopRequestID: String, frame: WKFrameInfo) {
        guard let current = session else {
            respond(id: stopRequestID, values: ["ok": false, "error": "recognition_not_active"], frame: frame)
            return
        }
        current.stopping = true
        speech.stopListening()
        respond(id: stopRequestID, values: ["ok": true, "stopping": true], frame: frame)

        DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self, weak current] in
            guard let self, let current, self.session === current else { return }
            self.speech.cancel()
            self.promotePartial(current)
            if current.chunks.isEmpty { self.fail(current, error: "empty_transcript") }
            else { self.complete(current) }
        }
    }

    private func complete(_ current: Session) {
        guard session === current else { return }
        session = nil
        let text = current.chunks.joined(separator: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        if text.isEmpty {
            respond(id: current.id, values: ["ok": false, "error": "empty_transcript"], frame: current.frame)
        } else {
            respond(id: current.id, values: ["ok": true, "text": text], frame: current.frame)
        }
    }

    private func fail(_ current: Session, error: String) {
        guard session === current else { return }
        session = nil
        speech.cancel()
        respond(id: current.id, values: ["ok": false, "error": error], frame: current.frame)
    }

    private func isAllowed(_ origin: WKSecurityOrigin) -> Bool {
        guard allowedSchemes.contains(origin.protocol) else { return false }
        if origin.protocol == "file" { return true }
        return allowedHosts.contains(origin.host)
    }

    private func respond(id: String, values: [String: Any], frame: WKFrameInfo) {
        guard frame.isMainFrame, isAllowed(frame.securityOrigin) else { return }
        var payload = values
        payload["id"] = id
        guard JSONSerialization.isValidJSONObject(payload),
              let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else { return }

        DispatchQueue.main.async { [weak self, weak webView] in
            guard let self, let webView else { return }
            guard frame.isMainFrame, self.isAllowed(frame.securityOrigin) else { return }
            webView.evaluateJavaScript(
                "globalThis.__FP_NATIVE_SPEECH_BRIDGE_V1_DELIVER__(\(json))",
                in: frame,
                in: .page
            ) { _ in
                // Deliberately no fallback to the current frame.
                // Navigation invalidates the captured frame and drops the reply.
            }
        }
    }

    private static func errorCode(_ error: FamilyPilotOnDeviceSpeechV1.VoiceError) -> String {
        switch error {
        case .speechPermissionDenied: return "speech_permission_denied"
        case .microphonePermissionDenied: return "microphone_permission_denied"
        case .onDeviceRecognitionUnavailable: return "on_device_speech_unavailable"
        case .recognitionBusy: return "recognition_busy"
        case .recognitionFailed: return "speech_recognition_failed"
        case .emptyTranscript: return "empty_transcript"
        }
    }
}
